### ISSUE TITLE
Out of memory exception

### ISSUE THREAD

**Issue by ashesman** [V2GClarity/RISE-V2G#37]
Hi, I am a first time user of this project and new to V2G.  Thanks for giving access to such a useful project.  I was having a play with the SECC server application and discovered a situation where memory allocation fails.  

Line 125 of ConnectionHandler.java can try allocate a large memory block if the payload length is incorrect.  You can simulate this by opening a raw TCP connection (eg netcat) to the server and send some random characters.  I know this is not intended use but a possible scenario should the connection get out of sync or during development...

I have not looked at the protocol in detail yet to know if there are limitations on the length of a payload that could be checked earlier? [Edit] Table 9 of ISO15118-2 says payload length can be 0...4294967295.

The offending line:
```
getLogger().debug("Length of V2GTP payload in bytes according to V2GTP header: " + getPayloadLength());
setV2gTPPayload(new byte[getPayloadLength()]);
```

Example debug output:

```
2019-09-12T11:49:06,995 DEBUG [ConnectionThread fe80:0:0:0:e2d5:5eff:fe68:7a77%2] ConnectionHandler: Length of V2GTP payload in bytes according to V2GTP header: 1935959667
Exception in thread "ConnectionThread fe80:0:0:0:e2d5:5eff:fe68:7a77%2" java.lang.OutOfMemoryError: Java heap space
```
Regards

Ashley

**Comment by MarcMueltin** [comment]
Hi @ashesman,

Thanks for the hint, that's a good catch. The standard itself says that the payload shall not be bigger than 4 GB (as that is what you can encode in 4 bytes of the "payload length" fields of the header). Why it was decided to use 4 bytes for this ... I don't know.

However, an EXI-encoded ISO 15118 message is rarely bigger than a few dozen bytes, for sure less than 1 KB. So it is a good idea to check that header length first before trying to create a byte array of that size. So I'll address this in my next update and let you know once it is done.

Best, 
Marc

**Comment by AxelVoitier** [comment]
Hello,

In my own implementation I have a check in the header processing part that close immediately the connection if it is above a limit I defined much much lower than the 4 GB non-sense.

The standard would be better off requiring implementations to support a minimum size (or, at worst, several minimum sizes depending on the supported functionalities). I am sure none of the implementations running on much smaller targets can handle 4GB messages...

It hasn't become an issue in the field though (ie. never came to face a peer that cannot process large messages). For all the troubles in EXI coding, at least it compresses well :D.


Cheers,
Axel

**Comment by MarcMueltin** [comment]
Fixed with commit 4546e0e22921bf06bda49a0fc374cb6d556d1bf8
